File: backend/schemas/analytics.py
from typing import List, Optional, Dict
from pydantic import BaseModel


# ── Timeline ──────────────────────────────────────────────────
class TimelineBucket(BaseModel):
    period: str          # "2024-01" for month, "2024" for year, "2024-W03" for week
    label: str           # human-readable: "Jan 2024"
    commits: int
    insertions: int
    deletions: int
    unique_authors: int


class TimelineResponse(BaseModel):
    repository_id: int
    granularity: str     # "week" | "month" | "year"
    buckets: List[TimelineBucket]


# ── Hotspots ──────────────────────────────────────────────────
class HotspotFile(BaseModel):
    filepath: str
    filename: str
    extension: str
    directory: str
    change_count: int
    insertions: int
    deletions: int
    unique_authors: int
    risk_score: float
    risk_label: str
    risk_reason: str
    top_author: str
    top_author_pct: float

    model_config = {"from_attributes": True}


class HotspotsResponse(BaseModel):
    repository_id: int
    total_files: int
    high_risk_count: int
    medium_risk_count: int
    low_risk_count: int
    hotspots: List[HotspotFile]


# ── Ownership ──────────────────────────────────────────────────
class OwnerEntry(BaseModel):
    author: str
    email: str
    commit_count: int
    percentage: float


class ModuleOwnership(BaseModel):
    module: str           # directory path e.g. "src/auth"
    total_commits: int
    owners: List[OwnerEntry]  # sorted by percentage desc


class OwnershipResponse(BaseModel):
    repository_id: int
    modules: List[ModuleOwnership]


# ── Health ────────────────────────────────────────────────────
class HealthScoreBreakdown(BaseModel):
    label: str
    score: float
    weight: float
    description: str


class HealthResponse(BaseModel):
    repository_id: int
    health_score: float
    grade: str           # A / B / C / D / F
    breakdown: List[HealthScoreBreakdown]
    summary: str


# ── Milestones ─────────────────────────────────────────────────
class Milestone(BaseModel):
    year: int
    month: Optional[int]
    event: str
    type: str            # birth | release | feature | infra | devops | perf
    commit_sha: Optional[str]
    commit_message: Optional[str]


class MilestonesResponse(BaseModel):
    repository_id: int
    milestones: List[Milestone]

File: frontend/src/api/analytics.ts
import { apiClient } from './client'
import type {
  ContributorsResponse,
  TimelineResponse,
  MilestonesResponse,
  HotspotsResponse,
  OwnershipResponse,
  HealthResponse,
  Granularity,
  RiskLabel,
} from '@/types/api'

export const analyticsApi = {
  contributors: async (repoId: number, limit = 20): Promise<ContributorsResponse> => {
    const { data } = await apiClient.get<ContributorsResponse>(
      `/repositories/${repoId}/contributors`,
      { params: { limit } }
    )
    return data
  },

  timeline: async (repoId: number, granularity: Granularity = 'month'): Promise<TimelineResponse> => {
    const { data } = await apiClient.get<TimelineResponse>(
      `/repositories/${repoId}/timeline`,
      { params: { granularity } }
    )
    return data
  },

  milestones: async (repoId: number): Promise<MilestonesResponse> => {
    const { data } = await apiClient.get<MilestonesResponse>(
      `/repositories/${repoId}/milestones`
    )
    return data
  },

  hotspots: async (repoId: number, risk?: RiskLabel, limit = 50): Promise<HotspotsResponse> => {
    const { data } = await apiClient.get<HotspotsResponse>(
      `/repositories/${repoId}/hotspots`,
      { params: { risk, limit } }
    )
    return data
  },

  ownership: async (repoId: number): Promise<OwnershipResponse> => {
    const { data } = await apiClient.get<OwnershipResponse>(
      `/repositories/${repoId}/ownership`
    )
    return data
  },

  health: async (repoId: number): Promise<HealthResponse> => {
    const { data } = await apiClient.get<HealthResponse>(
      `/repositories/${repoId}/health`
    )
    return data
  },
}
